core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java

```diff
@@ -318,4 +424,48 @@ public synchronized T get()
             }
         }
     }
+
+    private static Set<Long> createSetForLongJavaType(Set<Long> values, Type type, MethodHandle hashCodeHandle, MethodHandle equalsHandle)
+    {
+        if (values.isEmpty() || !canSkipTypeSpecificEqualsAndHashCode(type)) {
+            return new LongOpenCustomHashSet(values, 0.25f, new LongStrategy(hashCodeHandle, equalsHandle, type));
+        }
+
+        Iterator<Long> iterator = values.iterator();
+        long min = iterator.next();
+        long max = min;
+        while (iterator.hasNext()) {
+            long value = iterator.next();
+            min = min(min, value);
+            max = max(max, value);
+        }
+
+        // A Set based on a fastutil open hashset uses (next power of two of size / load-factor)
+        // slots in a hash table which is an array of longs.
+        long fastUtilSetEntries = (int) min(1 << 30, nextPowerOfTwo((long) ceil(values.size() / 0.25f)));
+        BigInteger range = BigInteger.valueOf(max)
+                .subtract(BigInteger.valueOf(min))
+                .add(BigInteger.valueOf(1));
+        // A Set based on a bitmap uses (max - min) / 64 longs
+        // Create a bitset only if it uses fewer entries than the equivalent hash set
+        if (range.compareTo(BigInteger.valueOf(Integer.MAX_VALUE)) > 0
+                || (range.intValueExact() / 64) + 1 > fastUtilSetEntries) {
+            return new LongOpenHashSet(values, 0.25f);
+        }
+        return new LongBitSetFilter(values, min, max);
+    }
+
+    private static boolean canSkipTypeSpecificEqualsAndHashCode(Type type)
+    {
+        // Types for which we can safely use equality and hashCode on the stored long value
```

Comment:
Why going through the Type operators is worse?

Reply:
Going through the LongStrategy wrapper makes LongOpenCustomHashSet slower than LongOpenHashSet (e.g. ~400 ops/sec vs ~600 ops/sec).
I also use this method to ensure that using the bitset representation is safe as that relies on operating on the underlying long value directly, let me rename method to be a bit clearer.